tower-http/src/trace.rs

```diff
@@ -0,0 +1,230 @@
+use crate::LatencyUnit;
+use crate::{GetTraceStatus, GetTraceStatusFromHttpStatus, TraceStatus};
+use futures_util::ready;
+use http::{Request, Response};
+use pin_project::pin_project;
+use std::future::Future;
+use std::time::Instant;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+use tracing::{
+    field::{debug, display},
+    Level, Span,
+};
+
+#[derive(Clone, Debug)]
+pub struct TraceLayer<T> {
+    record_headers: bool,
+    span: Option<Span>,
+    latency_unit: LatencyUnit,
+    get_trace_status: T,
+    record_full_uri: bool,
+}
+
+impl Default for TraceLayer<GetTraceStatusFromHttpStatus> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TraceLayer<GetTraceStatusFromHttpStatus> {
+    pub fn new() -> Self {
+        Self {
+            record_headers: false,
+            span: None,
+            latency_unit: LatencyUnit::Millis,
+            get_trace_status: GetTraceStatusFromHttpStatus(()),
+            record_full_uri: false,
+        }
+    }
+}
+
+impl<T> TraceLayer<T> {
+    pub fn record_headers(mut self, record_headers: bool) -> Self {
+        self.record_headers = record_headers;
+        self
+    }
+
+    /// Provide a custom span. The span is expected to at least have the fields `method`, `path`,
+    /// and `headers`. If any of the fields are missing from the span they'll also be missing from
+    /// whatever output you may have configured.
+    ///
+    /// The default span uses `DEBUG` level and is called `http-request`.
+    pub fn span(mut self, span: Span) -> Self {
+        self.span = Some(span);
+        self
+    }
+
+    pub fn latency_unit(mut self, latency_unit: LatencyUnit) -> Self {
+        self.latency_unit = latency_unit;
+        self
+    }
+
+    pub fn record_full_uri(mut self, record_full_uri: bool) -> Self {
+        self.record_full_uri = record_full_uri;
+        self
+    }
+
+    pub fn get_trace_status<K>(self, get_trace_status: K) -> TraceLayer<K> {
```

Comment:
same naming quibble as above. also, why `K`?

Reply:
No particular reason. Its called `T` on the `impl` and `K` comes after `T` 😅